baram/ec2_manager.py

```diff
@@ -1,105 +1,274 @@
+        :return: GroupIds
+        """
+        # 1. Get security groups not related to anything in ec2
+        sgs = self.list_sgs()
+        valid_sg_ids = set([pair['sg_id'] for pair in self.list_vpc_sg_eni_subnets()])
+        result = [sg['GroupId'] for sg in sgs
+                  if sg['GroupId'] not in valid_sg_ids
+                  or (description_filter in sg['Description']
+                      and sum([domain_id in sg['Description'] for domain_id in unused_domain_ids]) == 0)]
+
+        return set(result)
+
+    def list_vpc_sg_eni_subnets(self):
+        """
+        Return the list of all pairs of {VpcId, GroupId, NetworkInterfaceId, SubnetId}.
+
+        :return: The list of {vpc_id, security_group_id, eni_id, subnet_id}
+        """
+        vpcs = [vpc for vpc in self.list_vpcs() if vpc['State'] == 'available']
+        vpc_ids = [vpc['VpcId'] for vpc in vpcs]
+
+        result = []
+        for vpc_id in vpc_ids:
+            specified_sg_ids = self.get_sg_ids_with_vpc_id(vpc_id=vpc_id)
+            for sg_id in specified_sg_ids:
+                specified_enis = self.get_eni_with_sg_id(sg_id=sg_id)
+                pairs = {'vpc_id': vpc_id, 'sg_id': sg_id}
+                if len(specified_enis) != 0:
+                    for eni in specified_enis:
+                        pairs['eni_id'] = eni['NetworkInterfaceId']
+                        pairs['subnet_id'] = eni['SubnetId']
+                result.append(pairs)
+
+        return result
+
+    def get_sg_ids_with_vpc_id(self, vpc_id: str):
+        """
+        Get security group ids of specific vpc.
+
+        :param vpc_id: VpcId
+        :return: GroupId
+        """
+        sgs = self.cli.describe_security_groups()['SecurityGroups']
+        return [sg['GroupId'] for sg in sgs if sg['VpcId'] == vpc_id]
+
+    def get_eni_with_sg_id(self, sg_id: str):
+        """
+        Get network interfaces with specific security group.
+
+        :param sg_id: GroupId
+        :return: NetworkInterfaces
+        """
+        enis = self.cli.describe_network_interfaces()['NetworkInterfaces']
+        return [eni for eni in enis if eni['Groups'] != [] and sg_id in [x['GroupId'] for x in eni['Groups']]]
+
+    def list_sg_relations(self):
+        """
+        Describe security groups with related security groups, including egrss and ingress status.
+
+        :return: SecurityGroupId, IsEgress, RelatedSecurityGroupId
+        """
+        sg_rules = self.cli.describe_security_group_rules()['SecurityGroupRules']
+
+        sg_ids = [sg_rule['GroupId'] for sg_rule in sg_rules]
+        is_obs = [sg_rule['IsEgress'] for sg_rule in sg_rules]
+        referenced_sgs = [x['ReferencedGroupInfo']['GroupId']
+                          if 'ReferencedGroupInfo' in x.keys() else None for x in sg_rules]
+
+        result = []
+        for i in range(len(sg_rules)):
+            result.append({'sg_id': sg_ids[i],
+                           'is_ob': is_obs[i],
```

Comment:
it's better to replace as egress or ingress. It's more natural and frequent term in aws env.

Reply:
I'll change all ib & ob to ingress & egress.